vespajlib/src/main/java/ai/vespa/llm/client/openai/OpenAiClient.java

```diff
@@ -67,44 +76,101 @@ public List<Completion> complete(Prompt prompt, InferenceParameters options) {
-                            }
-                        }
-                    });
-                } catch (Exception e) {
-                    completionFuture.completeExceptionally(e);
-                }
-            }).exceptionally(e -> {
-                completionFuture.completeExceptionally(e);
+                handleHttpResponse(response, context);
+            }).exceptionally(exception -> {
+                handleHttpException(exception, context, attempt);
                 return null;
             });
-            return completionFuture;
 
         } catch (Exception e) {
             throw new RuntimeException(e);
         }
     }
 
+    private void handleHttpResponse(HttpResponse<Stream<String>> response, CompletionContext context) {
+        try {
+            int responseCode = response.statusCode();
+            if (responseCode != 200) {
+                throw new LanguageModelException(responseCode, response.body().collect(Collectors.joining()));
+            }
+            try (Stream<String> lines = response.body()) {
+                lines.forEach(line -> processLine(context, line));
+            }
+        } catch (Exception e) {
+            context.completionFuture().completeExceptionally(e);
+        }
+    }
+
+    private void processLine(CompletionContext context, String line) {
+        if (line.startsWith(DATA_FIELD)) {
+            var root = SlimeUtils.jsonToSlime(line.substring(DATA_FIELD.length())).get();
+            var completion = toCompletions(root, "delta").get(0);
+            context.consumer().accept(completion);
+            if (!completion.finishReason().equals(Completion.FinishReason.none)) {
+                context.completionFuture().complete(completion.finishReason());
+            }
+        }
+    }
+
+    private void processLines(CompletionContext context, Stream<String> lines) {
+    }
+
+    private void waitBeforeRetry() {
+        try {
+            TimeUnit.MILLISECONDS.sleep(RETRY_DELAY_MS);
+        } catch (InterruptedException ie) {
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    private boolean shouldRetry(Throwable exception) {
+        Throwable cause = exception.getCause();
+        if (cause instanceof SocketException && cause.getMessage().contains("Connection reset")) {
+            return true;
+        }
+        if (cause instanceof HttpConnectTimeoutException) {
+            return true;
+        }
+        if (cause instanceof HttpTimeoutException) {
+            return true;
+        }
+        return false;
+    }
+
+    private void handleHttpException(Throwable exception, CompletionContext context, int attempt) {
+        if (shouldRetry(exception)) {
+            if (attempt < MAX_RETRIES) {
+                waitBeforeRetry();
+                completeAsyncAttempt(context, attempt + 1);
+            } else {
+                context.completionFuture().completeExceptionally(new RuntimeException("OpenAI: max retries reached"));
```

Comment:
Consider including the original exception as cause to the new `RuntimeException` instance.

Reply:
Thanks!